src/xrpld/app/tx/detail/DepositPreauth.cpp

```diff
@@ -18,15 +18,285 @@
 //==============================================================================
 
 #include <xrpld/app/tx/detail/DepositPreauth.h>
+#include <xrpld/app/tx/detail/OnChainCredentials.h>
 #include <xrpld/ledger/View.h>
 #include <xrpl/basics/Log.h>
 #include <xrpl/protocol/Feature.h>
 #include <xrpl/protocol/Indexes.h>
 #include <xrpl/protocol/TxFlags.h>
+#include <xrpl/protocol/digest.h>
 #include <xrpl/protocol/st.h>
 
+#include <optional>
+#include <unordered_set>
+
 namespace ripple {
 
+static std::optional<std::reference_wrapper<STArray const>>
+getOptArray(STTx const& tx, SField const& field)
+{
+    int index = tx.getFieldIndex(field);
+    if (index < 0)
+        return std::nullopt;
+    STBase const& f = tx.peekAtIndex(index);
+    if (f.getSType() != STI_ARRAY)
+        return std::nullopt;
+    return std::cref(static_cast<STArray const&>(f));
+}
+
+static AccountID const&
+getAcount(STObject const& o, SField const& field)
+{
+    static AccountID const emtpy;
+    int index = o.getFieldIndex(field);
+    if (index < 0)
+        return emtpy;
+    STBase const& f = o.peekAtIndex(index);
+    if (f.getSType() != STI_ACCOUNT)
+        return emtpy;
+    return static_cast<STAccount const&>(f).value();
+}
+
+static Slice
+getCredType(STObject const& o, SField const& field)
+{
+    static Slice const emtpy;
+    int index = o.getFieldIndex(field);
+    if (index < 0)
+        return emtpy;
+    STBase const& f = o.peekAtIndex(index);
+    if (f.getSType() != STI_VL)
+        return emtpy;
+    return static_cast<STBlob const&>(f).value();
+}
+
+static std::optional<std::reference_wrapper<STVector256 const>>
+getOptVector(STTx const& tx, SField const& field)
+{
+    int index = tx.getFieldIndex(field);
+    if (index < 0)
+        return std::nullopt;
+    STBase const& f = tx.peekAtIndex(index);
+    if (f.getSType() != STI_VECTOR256)
+        return std::nullopt;
+    return std::cref(static_cast<STVector256 const&>(f));
+}
+
+static uint256
+hashAuthCred(STObject const& o)
+{
+    sha512_half_hasher h;
+    {
+        STBase const& f = o.peekAtField(sfIssuer);
+        AccountID const& v(static_cast<STAccount const&>(f).value());
+        hash_append(h, v);
+    }
+    {
+        STBase const& f = o.peekAtField(sfCredentialType);
+        Slice const& v(static_cast<STBlob const&>(f).value());
+        hash_append(h, v);
+    }
+    uint256 const x(h);
+    return x;
+}
+
+struct HashUint256
+{
+    std::size_t
+    operator()(uint256 const& x) const noexcept
+    {
+        std::size_t u = 0;
+        std::size_t const* const b =
+            reinterpret_cast<std::size_t const*>(x.data());
+        std::size_t const* const e = b + 256 / 8 / sizeof(std::size_t);
+
+        for (std::size_t const* c = b; c < e; ++c)
+            boost::hash_combine(u, *c);
+
+        return u;
+    }
+};
+
+NotTEC
+DepositPreauth::preflightCheckCredentials(
```

Comment:
This function should be renamed to be more general.
```suggestion
DepositPreauth::preflightCheck(
```

Reply:
fixed